lib/iterator.js

```diff
@@ -0,0 +1,317 @@
+  }
+
+  filter(predicate, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FilterIterator(this, predicate, thisArg);
+  }
+
+  flat(depth = 1) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatIterator(this, depth);
+  }
+
+  flatMap(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatMapIterator(this, mapper, thisArg);
+  }
+
+  zip(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new ZipIterator(this, iterators);
+  }
+
+  join(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new JoinIterator(this, iterators);
+  }
+
+  take(amount) {
+    // eslint-disable-next-line no-use-before-define
+    return new TakeIterator(this, amount);
+  }
+
+  skip(amount) {
+    for (let i = 0; i < amount; i++) {
+      this.next();
+    }
+    return this;
+  }
+}
+
+class MapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    const { done, value } = this.base.next();
+    return {
+      done,
+      value: done ? undefined : this.mapper.call(this.thisArg, value),
+    };
+  }
+}
+
+class FilterIterator extends Iterator {
+  constructor(base, predicate, thisArg) {
+    super(base);
+    this.predicate = predicate;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    for (const value of this.base) {
+      if (this.predicate.call(this.thisArg, value)) {
+        return { done: false, value };
+      }
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatIterator extends Iterator {
+  constructor(base, depth) {
+    super(base);
+    this.depth = 0;
+    this.stack = new Array(depth + 1);
+    this.stack[0] = base;
+  }
+
+  next() {
+    while (this.depth >= 0) {
+      const top = this.stack[this.depth];
+      const next = top.next();
+
+      if (next.done) {
+        this.stack[this.depth] = undefined;
+        this.depth--;
+        continue;
+      }
+
+      if (this.depth === this.stack.length - 1) {
+        return next;
+      }
+
+      if (next.value[Symbol.iterator]) {
+        this.stack[++this.depth] = next.value[Symbol.iterator]();
+        continue;
+      }
+
+      if (!next.value.next) {
+        return next;
+      }
+
+      this.stack[++this.depth] = next.value[Symbol.iterator]();
+    }
+
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatMapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+    this.currentIterator = undefined;
+  }
+
+  next() {
+    if (!this.currentIterator) {
+      const next = this.base.next();
+      if (next.done) {
+        return next;
+      }
+
+      const value = this.mapper.call(this.thisArg, next.value);
+      if (!value[Symbol.iterator] && !value.next) {
```

Comment:
But frankly I'd just drop support for iterators that are not iterables, because I can't see any valid use case for them, but I do perceive non-obvious bugs in application code. Can you please give an example when they might be useful?